### PR TITLE
Development stats cache and validations

### PR DESCRIPTION
**Pull request by waleedhammam**
#### [Commit 1](https://github.com/threefoldfoundation/tft-stellar/commit/5961d554345300cff1563959d62500aa608d5b45): 

- Make a service to prefill tft & tfta stats every 1 hour
- Drop network param and stick to public
- Add collection time to know the age of the cache
- If a request comes in while the cache has not been filled yet, respond with a 503 Service Unavailable
- Stick to only the detailed statistics for TFT and TFTA need to be calculated

#### [Commit 2](https://github.com/threefoldfoundation/tft-stellar/commit/ef6b52e38cde4275ad1d9e9109e093c0e466268b): 

- Add the rest of endpoint cache requests in the service aswell to be every 1hour
- Move out the all the helper function to a SAL included in the package
- If a request comes in while the cache has not been filled yet, respond with a 503 Service Unavailable

#### [Commit 3](https://github.com/threefoldfoundation/tft-stellar/pull/402/commits/01a41ac042979b2338b84979678bc650898ac6b5):

- Add sanity checks for illegal values

### Related  Issues

- https://github.com/threefoldfoundation/tft-stellar/issues/401
- https://github.com/threefoldfoundation/tft-stellar/issues/360
- https://github.com/threefoldfoundation/tft-stellar/issues/366
- https://github.com/threefoldfoundation/tft-stellar/issues/364
- https://github.com/threefoldfoundation/tft-stellar/issues/361
- https://github.com/threefoldfoundation/tft-stellar/issues/365

### Screenshots

Screenshot from 2021-06-22 14-50-06

Screenshot from 2021-06-22 14-55-15

Screenshot from 2021-06-22 14-55-28

Screenshot from 2021-06-22 14-55-42

Screenshot from 2021-06-22 15-11-45

### CODE REVIEW

File: ThreeBotPackages/tft_statistics/services/updatestats.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    def job(self):
+        try:
+            # update foundation wallets
+            j.logger.info(f"Updating foundation wallets")
+            foundation_wallets = update_foundation_wallets_data()
+            j.logger.debug(f"Foundation wallets updated successfully {foundation_wallets}")
+
+            # update total tft & tfta total tokens
+            j.logger.info("Updating total TFT")
+            total_tft = update_total_tft(tokencode="TFT")
```
Comment (robvanmieghem) [inline ThreeBotPackages/tft_statistics/services/updatestats.py:35]: Can't we derive this from the the detailed tft statistics which are calculated later?

File: ThreeBotPackages/tft_statistics/services/updatestats.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            j.logger.debug(f"Successfully updated total tft with: {total_tft}")
+
+            j.logger.info("Updating total TFTA")
+            total_tft = update_total_tft(tokencode="TFTA")
```
Comment (robvanmieghem) [inline ThreeBotPackages/tft_statistics/services/updatestats.py:39]: same remark

File: ThreeBotPackages/tft_statistics/services/updatestats.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            j.logger.debug(f"Successfully updated total tft with: {total_tft}")
+
+            # update total tft & tfta unlocked tokens
+            j.logger.info("Updating total unlocked TFT")
+            total_unlocked_tft = update_total_unlocked_tft(tokencode="TFT")
```
Comment (robvanmieghem) [inline ThreeBotPackages/tft_statistics/services/updatestats.py:44]: same remark

File: ThreeBotPackages/tft_statistics/sals/stats_sals.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    total_vested_tft = stats["total_vested"]
+    total_foundation = stats["total_foundation"]
+
+    total_unlocked_tft = total_tft - total_locked_tft - total_vested_tft - total_foundation
+
+    redis.set(f"{tokencode}-total_unlocked_tft", total_unlocked_tft)
+    return total_unlocked_tft
+
+
+def update_stats(tokencode: str = "TFT", detailed: bool = True) -> dict:
```
Comment (robvanmieghem) [inline ThreeBotPackages/tft_statistics/sals/stats_sals.py:91]: not detailed is no longer needed

**Review by robvanmieghem — CHANGES_REQUESTED** [review]
Looks good Waleed, thanks, only a few small remarks

File: ThreeBotPackages/tft_statistics/services/updatestats.py (hunk 1 above)
Comment (waleedhammam) [reply ThreeBotPackages/tft_statistics/services/updatestats.py]: done

File: ThreeBotPackages/tft_statistics/services/updatestats.py (hunk 2 above)
Comment (waleedhammam) [reply ThreeBotPackages/tft_statistics/services/updatestats.py]: done

File: ThreeBotPackages/tft_statistics/services/updatestats.py (hunk 3 above)
Comment (waleedhammam) [reply ThreeBotPackages/tft_statistics/services/updatestats.py]: done

File: ThreeBotPackages/tft_statistics/sals/stats_sals.py (hunk 4 above)
Comment (waleedhammam) [reply ThreeBotPackages/tft_statistics/sals/stats_sals.py]: done

**Review by robvanmieghem — APPROVED** [review]
thank you

**Pull request merged**